### ISSUE TITLE
Some thoughts on structural loss

### ISSUE THREAD

**Issue by zhaoxin111** [cfzd/Ultra-Fast-Lane-Detection#203]
First of all, it's an excellent job. Thank you for open source.

As the author mentioned in the paper, structural loss is an important innovation.

However, the loss of structure may cause the instability of training. I think this part can only calculate the structure loss of GT part, which can avoid background prediction interfering with the training.
![image](https://user-images.githubusercontent.com/9295202/116500094-670d5600-a8e0-11eb-9930-11bca2ffc99e.png)

**Comment by cfzd** [comment]
@zhaoxin111 
Excellent idea! I think this makes sense for more stable training!

**Comment by zhaoxin111** [comment]
Limiting the structure loss to gt and positive sample prediction can more effectively alleviate unstable training. At the same time, it is found that reducing the weight of structure loss is better in testing on your own data set.
![image](https://user-images.githubusercontent.com/9295202/117231118-2322cf80-ae51-11eb-8781-790426911ce4.png)

**Comment by zhaoxin111** [comment]
In addition to thinking about structural loss, I also found a problem. When the line to be trained is almost a vertical line, there will be a certain offset between gt loc and the true position of the line due to the quantization operation during the calculation of gt row loc. This problem can be alleviated by increasing num_grid or reducing row anchor, but it cannot be eradicated.

**Comment by eriche2016** [comment]
Hi, I found this [line](https://github.com/cfzd/Ultra-Fast-Lane-Detection/blob/master/utils/loss.py#L72) should be 
```
loss /= (len(diff_list1) - 1)

```
though it may not influence the result too much

**Comment by zhaoxin111** [comment]
> Hi, I found this [line](https://github.com/cfzd/Ultra-Fast-Lane-Detection/blob/master/utils/loss.py#L72) should be
> 
> ```
> loss /= (len(diff_list1) - 1)
> ```
> 
> though it may not influence the result too much

I have modified the loss calculation strategy, so -1 is not needed

**Comment by eriche2016** [comment]
> > Hi, I found this [line](https://github.com/cfzd/Ultra-Fast-Lane-Detection/blob/master/utils/loss.py#L72) should be
> > ```
> > loss /= (len(diff_list1) - 1)
> > ```
> > 
> > 
> >   
> >     
> >     
> > 
> >     
> >     
> > 
> >   
> > 
> > 
> > though it may not influence the result too much
> 
> I have modified the loss calculation strategy, so -1 is not needed

@zhaoxin111  Hi, dear friend. Thanks for opening this issue, and I also came across the problem of unstable training using the structure loss, which gives me nan loss. Here what I am referring to is the original loss calculation code by the authors.

I have read your comments carefully on proposing two modifications to the calculation of structure loss. 
The first version is reasonable, while the second version may also unstable at the beginning of the training phase, 
since the positive prediction may be unreliable, and I am not sure whether it will cause problems. 

BTW, have you tried them in your experiments, and what performance can you obtain on public benchmarks like CULane?

**Comment by zhaoxin111** [comment]
The prediction in the initial training stage may indeed be unreliable, but the second version is more rigorous in the loss calculation than the first version. The second-order differential is calculated only when gt=1 & pred=1, so theoretically speaking, the initial stage of training The structure loss has a small effect, but as the accuracy of the model prediction increases, the structure loss will gradually become larger and optimize the convergence direction of the model during the training process and force the model prediction to trend toward a straight line. However, it should be noted that the weight of the structure loss  cannot be large. If it is found that when the classification loss cannot effectively converge, try to reduce the structural loss weight.

Tests on my own data show that adding the structure loss of this version can improve the accuracy to a certain extent, but it should be noted that if the grid_num is set to a large value, the weight of the structure loss should be reduced as much as possible. In the training process, I did not add sim loss, because my row_anchor settings are relatively small, and the impact of row_anchor settings has also been discussed earlier. @eriche2016

**Comment by eriche2016** [comment]
Thanks for your reply. I will try them in my experiments.

**Comment by Khadaz** [comment]
x = torch.nn.functional.softmax(x[:,:dim-1,:,:],dim=1)
embedding = torch.Tensor(np.arange(dim-1)).float().to(x.device).view(1,-1,1,1)

May I alter the "dim - 1" to "dim - 2" to achieve the same effect?
